File: DialedInCoffee/Presentation/Components/MachineSettingsEditor.swift
import SwiftUI

public struct MachineSettingsEditor: View {
    @Binding public var settings: MachineSettings
    public var showsMilk: Bool

    public init(settings: Binding<MachineSettings>, showsMilk: Bool = true) {
        self._settings = settings
        self.showsMilk = showsMilk
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Theme.Spacing.md) {
            row(label: "Grinder", value: settings.grinder, range: MachineRanges.grinderRange) { settings.grinder = $0 }
            row(label: "Strength", value: settings.strength, range: MachineRanges.strengthRange) { settings.strength = $0 }
            row(label: "Volume (ml)", value: settings.volumeML, range: MachineRanges.volumeRange, step: 5) { settings.volumeML = $0 }
            if showsMilk {
                row(label: "Milk (s)", value: settings.milkSeconds, range: MachineRanges.milkDurationRange) { settings.milkSeconds = $0 }
            }
            HStack {
                Text("Temperature")
                    .frame(width: 110, alignment: .leading)
                Picker("", selection: $settings.temperature) {
                    ForEach(TemperatureLevel.allCases) { level in
                        Text(level.displayName).tag(level)
                    }
                }
                .labelsHidden()
                .pickerStyle(.segmented)
            }
        }
    }

    @ViewBuilder
    private func row(
        label: String,
        value: Int,
        range: ClosedRange<Int>,
        step: Int = 1,
        set: @escaping (Int) -> Void
    ) -> some View {
        HStack {
            Text(label)
                .frame(width: 110, alignment: .leading)
            Stepper(value: Binding(get: { value }, set: { set($0) }), in: range, step: step) {
                Text("\(value)")
                    .monospacedDigit()
                    .frame(minWidth: 36, alignment: .leading)
            }
        }
    }
}

#Preview {
    StatefulPreviewWrapper(MachineSettings()) { binding in
        MachineSettingsEditor(settings: binding)
            .padding()
            .frame(width: 360)
    }
}

private struct StatefulPreviewWrapper<Value, Content: View>: View {
    @State private var value: Value
    private let content: (Binding<Value>) -> Content

    init(_ initial: Value, @ViewBuilder content: @escaping (Binding<Value>) -> Content) {
        self._value = State(initialValue: initial)
        self.content = content
    }

    var body: some View { content($value) }
}

File: android/app/src/main/kotlin/com/cawfee/ui/components/MachineSettingsEditor.kt
package com.cawfee.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.cawfee.domain.model.MachineRanges
import com.cawfee.domain.model.MachineSettings
import com.cawfee.domain.model.TemperatureLevel

/**
 * Editor for a [MachineSettings] value. Compose analogue of MachineSettingsEditor.swift —
 * grinder / strength / volume / (milk) sliders plus a temperature segmented control. Every
 * change re-clamps via the MachineSettings factory.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MachineSettingsEditor(
    settings: MachineSettings,
    onChange: (MachineSettings) -> Unit,
    modifier: Modifier = Modifier,
    showsMilk: Boolean = true,
) {
    Column(modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        IntSlider(
            label = "Grinder", value = settings.grinder, range = MachineRanges.grinderRange,
            suffix = "/ 7",
        ) { onChange(settings.with(grinder = it)) }

        IntSlider(
            label = "Strength", value = settings.strength, range = MachineRanges.strengthRange,
            suffix = "/ 10",
        ) { onChange(settings.with(strength = it)) }

        IntSlider(
            label = "Volume", value = settings.volumeML, range = MachineRanges.volumeRange,
            step = 5, suffix = "ml",
        ) { onChange(settings.with(volumeML = it)) }

        if (showsMilk) {
            IntSlider(
                label = "Milk", value = settings.milkSeconds, range = MachineRanges.milkDurationRange,
                suffix = "s",
            ) { onChange(settings.with(milkSeconds = it)) }
        }

        Text("Temperature", fontWeight = FontWeight.Medium)
        SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
            TemperatureLevel.entries.forEachIndexed { i, level ->
                SegmentedButton(
                    selected = settings.temperature == level,
                    onClick = { onChange(settings.with(temperature = level)) },
                    shape = SegmentedButtonDefaults.itemShape(i, TemperatureLevel.entries.size),
                ) { Text(level.displayName) }
            }
        }
    }
}

@Composable
private fun IntSlider(
    label: String,
    value: Int,
    range: IntRange,
    step: Int = 1,
    suffix: String = "",
    onChange: (Int) -> Unit,
) {
    Column {
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Text(label, fontWeight = FontWeight.Medium)
            Text("$value $suffix".trim())
        }
        val steps = ((range.last - range.first) / step) - 1
        Slider(
            value = value.toFloat(),
            onValueChange = { onChange(Math.round(it)) },
            valueRange = range.first.toFloat()..range.last.toFloat(),
            steps = if (steps > 0) steps else 0,
        )
    }
}
